graphqlapi/developer_docs.md

```diff
@@ -0,0 +1,299 @@
+example, use a different table name than the inferred default). They are defined in [CqlDirectives],
+and referenced throughout the model-building code. You can also download a text version from a
+running Stargate instance at `graphql-files/cql_directives.graphql`.
+
+#### Deployed (`/graphql/<keyspace>`)
+
+Once a user has deployed their own GraphQL schema, it replaces the CQL-first schema for that
+keyspace. As already mentioned, [GraphqlCache] contains logic to determine which variant to load.
+
+There are two ways that a GraphQL-first `GraphQL` instance can be created:
+
+* if the deploy operation just happened in this Stargate process, then [DeploySchemaFetcherBase]
+  already has the `GraphQL`, and puts it directly in the cache with `GraphqlCache.putDml`.
+* if the deployment happened via another Stargate instance, or if Stargate just restarted, we need
+  to reload the schema from `stargate_graphql.schema_source`. This uses a simplified version of the
+  deployment process described in the previous section, see
+  `GraphqlCache.LazySchemaFirstGraphqlHolder`.
+
+Unlike CQL-first, GraphQL-first does not react to external CQL schema changes: the user is supposed
+to have full control over the GraphQL schema, so we can't just change it behind their back. So our
+assumption is that once users go GraphQL-first, they will evolve their CQL schema via successive
+deploy operations, not by directly altering the database. If things get out of sync, there are no
+guarantees: queries might start to fail, or a saved schema might fail to reload.
+
+The data fetchers are in the package [graphqlfirst.fetchers.deployed]. They rely on the
+representation that was built by [MappingModel]. For example, the query fetcher relies on a query
+model that defines the target table, which parameters of the GraphQL operation will be mapped to
+`WHERE` clauses, etc. This part of the code is quite tedious because our mapping rules are very
+flexible: there are many different ways to define operations and map the results. The best way to
+get familiar with a fetcher is to look at the integration tests and trace their execution.
+
+### Batching
+
+Batching is a cross-cutting concern, both CQL-first and GraphQL-first support the `@atomic`
+directive to indicate that a set of mutations must be executed as a single CQL batch:
+
+```graphql
+mutation @atomic {
+  ... // mutations
+}
+```
+
+[StargateGraphqlContext] (via its inner class `BatchContext`) provides coordination services that
+allows independent fetchers to accumulate queries, and track which is the last one that must execute
+the batch.
+
+In the CQL-first code, [MutationFetcher] and [BulkMutationFetcher] are equipped to detect the
+directive and use the batch context. In the GraphQL-first code, there is a single
+[MutationFetcher][graphqlfirst.MutationFetcher] parent class that handles both regular and bulk
+operations. The GraphQL-first design is cleaner IMHO, so it would probably be a good idea to
```

Comment:
Should we create an issue and just remove this sentence?

Reply:
#1207